### PR TITLE
feat: add user group filter to the publish modal(Issue #433)

### PR DESCRIPTION
**Pull request by denys-kolomiitsev**
- Create reusable MultipleComboBox component;
- User groups filter with badges;
- Tests;

**Issue:** #433 

image

### CODE REVIEW

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Deployment status: success

File: src/components/Chat/PublishModal.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ import {
 import { onBlur } from '@/src/utils/app/style-helpers';
 
 import { ShareEntity } from '@/src/types/common';
-import { SharingType } from '@/src/types/share';
+import { SharingType, UserGoup } from '@/src/types/share';
```
Comment (IlyaBondar) [inline src/components/Chat/PublishModal.tsx]: typo -> `UserGroup`

File: src/components/Chat/PublishModal.tsx [hunk 2]
```diff
@@ -253,8 +256,16 @@ export default function PublishModal({
         <section className="flex grow flex-col gap-3">
           <h2>{t('Target Audience Filters')}</h2>
 
+          <CollapsableSection
+            name={'UserGroup'}
+            dataQa={`filter-${'UserGroup'}`}
```
Comment (IlyaBondar) [inline src/components/Chat/PublishModal.tsx]:
```suggestion
            dataQa="filter-user-group"
```

File: src/components/Chat/PublishModal.tsx [hunk 3]
```diff
@@ -253,8 +256,16 @@ export default function PublishModal({
         <section className="flex grow flex-col gap-3">
           <h2>{t('Target Audience Filters')}</h2>
 
+          <CollapsableSection
+            name={'UserGroup'}
```
Comment (IlyaBondar) [inline src/components/Chat/PublishModal.tsx]: translate name
```suggestion
            name={t('UserGroup')}
```

File: src/components/Chat/PublishModalUserGroupFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  { id: 'users-3', name: 'Users 3' },
+  { id: 'users-4', name: 'Users 4' },
+  { id: 'users-5', name: 'Users 5' },
+  { id: 'users-6', name: 'Users 6' },
+  { id: 'users-7', name: 'Users 7' },
+];
+
+//TODO remove from the file
+
+function getFilteredUserGroups<T>(
```
Comment (IlyaBondar) [inline src/components/Chat/PublishModalUserGroupFilter.tsx]: it looks like it generic method for any dropdown items and should be renamed in generic way and moved in util-file
```suggestion
function getFilteredItems<T>(
```

File: src/components/Chat/PublishModalUserGroupFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      (item) =>
+        !selectedItems.includes(item) &&
+        getItemLabel(item).toLowerCase().includes(lowerCasedInputValue),
+    );
+  }
+}
+// End of remove from file
+
+interface Props {
+  onChangeUserGoups: (userGroups: UserGoup[]) => void;
```
Comment (IlyaBondar) [inline src/components/Chat/PublishModalUserGroupFilter.tsx]: typo -> `onChangeUserGroups`
```suggestion
  onChangeUserGroups: (userGroups: UserGroup[]) => void;
```

File: src/components/Chat/PublishModal.tsx (hunk 1 above)
Comment (denys-kolomiitsev) [reply src/components/Chat/PublishModal.tsx]: Thank you! 
Fixed.

File: src/components/Chat/PublishModal.tsx (hunk 2 above)
Comment (denys-kolomiitsev) [reply src/components/Chat/PublishModal.tsx]: Fixed as suggested.

File: src/components/Chat/PublishModal.tsx (hunk 3 above)
Comment (denys-kolomiitsev) [reply src/components/Chat/PublishModal.tsx]: Translation added.

File: src/components/Chat/PublishModalUserGroupFilter.tsx (hunk 4 above)
Comment (denys-kolomiitsev) [reply src/components/Chat/PublishModalUserGroupFilter.tsx]: Name changed and function moved to the util-file.

File: src/components/Chat/PublishModalUserGroupFilter.tsx (hunk 5 above)
Comment (denys-kolomiitsev) [reply src/components/Chat/PublishModalUserGroupFilter.tsx]: Typo fixed.

File: src/components/Common/MultipleComboBox.tsx [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+  disabled?: boolean;
+  isHideDropdown?: boolean;
+  getItemLabel: (item: T) => string;
+  getItemValue: (item: T) => string;
+  getFilteredItems?: (
+    items: T[],
+    inputValue: string | undefined,
+    getItemLabel: (item: T) => string,
+    selectedItems?: T[],
+  ) => T[];
```
Comment (IlyaBondar) [inline src/components/Common/MultipleComboBox.tsx]: I don't think we need method as property (we have everything to execute it inside). 
image
Just execute `getFilteredItems` inside this component

Comment (denys-kolomiitsev) [reply src/components/Common/MultipleComboBox.tsx]: This method have been moved inside the compontent.

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Deployment status: success

File: src/components/Common/MultipleComboBox.tsx [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  if (!items) {
+    return !inputValue ||
+      selectedItems?.some((item) => getItemLabel(item) === inputValue)
+      ? []
+      : [inputValue as T];
+  }
+  if (!selectedItems) {
+    return items;
+  } else {
+    const lowerCasedInputValue = inputValue ? inputValue.toLowerCase() : '';
```
Comment (IlyaBondar) [inline src/components/Common/MultipleComboBox.tsx]:
```suggestion
    const lowerCasedInputValue = inputValue?.toLowerCase() || '';
```

File: src/components/Chat/PublishModalUserGroupFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { useCallback } from 'react';
```
Comment (IlyaBondar) [inline src/components/Chat/PublishModalUserGroupFilter.tsx]: I suggest to remove prefix `PublishModal` i.e. rename files:
- `PublishModalUserGroupFilter.tsx` -> `UserGroupFilter.tsx`
- `PublishModalTargetAudienceFilter.tsx` -> `TargetAudienceFilter.tsx`

Maybe these filters will be reused somewhere else

File: src/components/Common/MultipleComboBox.tsx (hunk 7 above)
Comment (denys-kolomiitsev) [reply src/components/Common/MultipleComboBox.tsx]: Changed as suggested.

File: src/components/Chat/PublishModalUserGroupFilter.tsx (hunk 8 above)
Comment (denys-kolomiitsev) [reply src/components/Chat/PublishModalUserGroupFilter.tsx]: Thank you for advising me!
Prefix removed.

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Deployment status: success

File: src/components/Common/MultipleComboBox.tsx [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+                  >
+                    {itemRow
+                      ? createElement(itemRow, { item })
+                      : getItemLabel(item)}
+                  </li>
+                ))
+              ) : (
+                <li className="px-3 py-2">
+                  {notFoundPlaceholder || t('No available items')}
+                </li>
```
Comment (IlyaBondar) [inline src/components/Common/MultipleComboBox.tsx]: I don't think we need to show it
image
I suggest to not show it if input is empty:
```suggestion
              ) : (inputValue &&
                <li className="px-3 py-2">
                  {notFoundPlaceholder || t('No available items')}
                </li>
```

File: src/components/Common/MultipleComboBox.tsx [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+                      <IconX height={12} width={12} />
+                    </span>
+                  </span>
+                );
+              })}
+
+            <input
+              disabled={disabled}
+              placeholder={placeholder || ''}
+              className="bg-transparent px-3 py-1 outline-none placeholder:text-secondary"
```
Comment (IlyaBondar) [inline src/components/Common/MultipleComboBox.tsx]: I think we need to limit input by width, also we don't need padding from right:
```suggestion
              className="max-w-[105px] bg-transparent pl-3 py-1 outline-none placeholder:text-secondary"
```
image

Also I thing that padding-left should be small if we have selectedItems:
```jsx
selectedItems.length ? 'pl-1' : 'pl-3'
```
image

**Comment by IlyaBondar** [comment]
/deploy-review

>Deployment status: success

File: src/components/Common/MultipleComboBox.tsx (hunk 9 above)
Comment (denys-kolomiitsev) [reply src/components/Common/MultipleComboBox.tsx]: Changed as suggested.

File: src/components/Common/MultipleComboBox.tsx (hunk 10 above)
Comment (denys-kolomiitsev) [reply src/components/Common/MultipleComboBox.tsx]: Thank you!
Styles changed as suggested.

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Deployment status: success

**Review by IlyaBondar — DISMISSED** [review]

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Deployment status: success

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Deployment status: success

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Deployment status: success

**Comment by denys-kolomiitsev** [comment]
/deploy-review

>Deployment status: success

**Review by IlyaBondar — APPROVED** [review]

**Pull request merged**